Rubberduck.Core/AutoComplete/Service/SelfClosingPairCompletionService.cs

```diff
@@ -0,0 +1,408 @@
+
+            return new CodeString(string.Join("\r\n", lines), finalCaretPosition,
+                new Selection(original.SnippetPosition.StartLine, 1, original.SnippetPosition.EndLine, 1));
+        }
+
+        private static Selection HandleBackspaceContinuations(string[] nonEmptyLines, Selection finalCaretPosition)
+        {
+            var lineIndex = Math.Min(finalCaretPosition.StartLine, nonEmptyLines.Length - 1);
+            var line = nonEmptyLines[lineIndex];
+            if (line.EndsWith(" _") && lineIndex == nonEmptyLines.Length - 1)
+            {
+                nonEmptyLines[lineIndex] = line.Remove(line.Length - 2);
+                line = nonEmptyLines[lineIndex];
+            }
+
+            if (lineIndex == nonEmptyLines.Length - 1)
+            {
+                line = nonEmptyLines[lineIndex];
+            }
+
+            if (line.EndsWith("&"))
+            {
+                // we're not concatenating anything anymore; remove concat operator too.
+                var concatOffset = line.EndsWith(" &") ? 2 : 1;
+                nonEmptyLines[lineIndex] = line.Remove(line.Length - concatOffset);
+            }
+            TrimNonEmptyLine(nonEmptyLines, lineIndex, "& vbNewLine");
+            TrimNonEmptyLine(nonEmptyLines, lineIndex, "& vbCrLf");
+            TrimNonEmptyLine(nonEmptyLines, lineIndex, "& vbCr");
+            TrimNonEmptyLine(nonEmptyLines, lineIndex, "& vbLf");
+
+            // we're keeping the closing quote, but let's put the caret inside:
+            line = nonEmptyLines[lineIndex];
+            var quoteOffset = line.EndsWith("\"") ? 1 : 0;
+            finalCaretPosition = new Selection(finalCaretPosition.StartLine, line.Length - quoteOffset);
+            return finalCaretPosition;
+        }
+
+        private static void TrimNonEmptyLine(string[] nonEmptyLines, int lineIndex, string ending)
+        {
+            var line = nonEmptyLines[lineIndex];
+            if (line.EndsWith(ending, StringComparison.OrdinalIgnoreCase))
+            {
+                var offset = line.EndsWith(" " + ending, StringComparison.OrdinalIgnoreCase)
+                    ? ending.Length + 1
+                    : ending.Length;
+                nonEmptyLines[lineIndex] = line.Remove(line.Length - offset);
+            }
+        }
+
+        private Selection FindMatchingTokenPosition(SelfClosingPair pair, CodeString original)
+        {
+            var code = string.Join("\r\n", original.Lines) + "\r\n";
+            code = code.EndsWith($"{pair.OpeningChar}{pair.ClosingChar}")
+                ? code.Substring(0, code.LastIndexOf(pair.ClosingChar) + 1)
+                : code;
+            var result = VBACodeStringParser.Parse(code, p => p.startRule());
+            if (((ParserRuleContext)result.parseTree).exception != null)
+            {
+                result = VBACodeStringParser.Parse(code, p => p.mainBlockStmt());
```

Comment:
`IStringParser` signpost

Reply:
In the current form, the `IStringParser` registered in the IoC container will not work here because it always expects an entire module.

I am currently pondering how to make the setup more flexible without braking the encapsulation.